### PR TITLE
Regression methods

### PR DESCRIPTION
**Pull request by Quid37**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):
Implementation of basic machine learning methods (stochastic linear regression and logistic regression) using aggregate functions interface. Has different strategies for updating model weights (simple gradient descent, momentum method, Nesterov method). Also supports mini-batches of custom size.
...

Detailed description (optional):

...

### CODE REVIEW

File: dbms/src/Columns/ColumnAggregateFunction.cpp [hunk 1]
```diff
@@ -65,23 +84,57 @@ MutableColumnPtr ColumnAggregateFunction::convertToValues() const
...
 
+MutableColumnPtr ColumnAggregateFunction::predictValues(Block & block, const ColumnNumbers & arguments) const
+{
+    MutableColumnPtr res;
+    if (convertion(&res))
+    {
+        return res;
+    }
+
+    auto ML_function_Linear = typeid_cast<AggregateFunctionMLMethod<LinearModelData, NameLinearRegression> *>(func.get());
```
Comment (hczhcz) [inline dbms/src/Columns/ColumnAggregateFunction.cpp]: It would be better to avoid hard-coding the function list

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+    {}
+
+    virtual ~IGradientComputer() = default;
+
+    /// Adds computed gradient in new point (weights, bias) to batch_gradient
+    virtual void compute(std::vector<Float64> * batch_gradient, const std::vector<Float64> &weights, Float64 bias,
+                         Float64 learning_rate, Float64 l2_reg_coef, Float64 target, const IColumn **columns, size_t row_num) = 0;
+
+    /// Now we should use predict_block function instead of predict
+    virtual void predict_block(ColumnVector<Float64>::Container &container,
```
Comment (KochetovNicolai) [inline dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: Just use `predict`

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+};
+
+
+class LinearRegression : public IGradientComputer
+{
+public:
+    LinearRegression()
+    {}
+
+    void compute(std::vector<Float64> * batch_gradient, const std::vector<Float64> &weights, Float64 bias,
```
Comment (KochetovNicolai) [inline dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: Why don't you pass batch_gradient by ref?

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+        {
+            derivative -= weights[i] * static_cast<const ColumnVector<Float64> &>(*columns[i]).getData()[row_num];
+        }
+        derivative *= (2 * learning_rate);
+
+        (*batch_gradient)[weights.size()] += derivative;
+        for (size_t i = 0; i < weights.size(); ++i)
+        {
+            (*batch_gradient)[i] +=
+                    derivative * static_cast<const ColumnVector<Float64> &>(*columns[i]).getData()[row_num]
```
Comment (KochetovNicolai) [inline dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: Save `static_cast<const ColumnVector<Float64> &>(*columns[i]).getData()[row_num]` into variable. Code will look better.

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+                        - 2 * l2_reg_coef * weights[i];
+        }
+    }
+
+    void predict_block(ColumnVector<Float64>::Container &container,
+                         Block &block,
+                         const ColumnNumbers &arguments,
+                         const std::vector<Float64> &weights, Float64 bias) const override
+    {
```
Comment (KochetovNicolai) [inline dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: Check `weights.size() + 1== arguments.size()`

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+        size_t rows_num = block.rows();
+        std::vector<Float64> results(rows_num, bias);
+
+        for (size_t i = 1; i < arguments.size(); ++i)
+        {
+            ColumnPtr cur_col = block.getByPosition(arguments[i]).column;
```
Comment (KochetovNicolai) [inline dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: safeGetByPosition

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+                const auto &element = (*cur_col)[row_num];
+                if (element.getType() != Field::Types::Float64)
+                    throw Exception("Prediction arguments must be values of type Float",
+                                    ErrorCodes::BAD_ARGUMENTS);
+
+                results[row_num] += weights[i - 1] * element.get<Float64>();
+            }
+        }
+
+        for (size_t row_num = 0; row_num != rows_num; ++row_num)
```
Comment (KochetovNicolai) [inline dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: Reserve `rows_num` rows in container

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+/**
+* IWeightsUpdater class defines the way to update current weights
+* and uses class GradientComputer on each iteration
+*/
+class IWeightsUpdater
+{
+public:
+    virtual ~IWeightsUpdater() = default;
+
+    virtual void add_to_batch(std::vector<Float64> * batch_gradient, std::shared_ptr<IGradientComputer> gc,
```
Comment (KochetovNicolai) [inline dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: Why not to pass gc by ref? (gc  - wrong name)

Comment (KochetovNicolai) [inline dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: Need comment for this method (and for all methods in this interface).

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+        weights_updater->read(buf);
+    }
+
+    void predict_block(ColumnVector<Float64>::Container &container, Block &block, const ColumnNumbers &arguments) const
+    {
+        gradient_computer->predict_block(container, block, arguments, weights, bias);
+    }
+
+private:
+    std::vector<Float64> weights;
```
Comment (KochetovNicolai) [inline dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: It would be better to allocate weights and gradient_batch on arena.

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+
+    Float64 learning_rate;
+    Float64 l2_reg_coef;
+    UInt32 batch_capacity;
+
+    UInt32 iter_num = 0;
+    std::vector<Float64> gradient_batch;
+    UInt32 batch_size;
+
+    std::shared_ptr<IGradientComputer> gradient_computer;
```
Comment (KochetovNicolai) [inline dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: I think, it would be better to avoid storing gradient_computer and weights_update by shared_ptr, but use GradientComputer and WeightsUpdater as part of aggregate function state (probably, inherit from them).
Thought, I am not sure we need it.

File: dbms/src/Columns/ColumnAggregateFunction.cpp [hunk 11]
```diff
@@ -33,6 +35,23 @@ void ColumnAggregateFunction::addArena(ArenaPtr arena_)
     arenas.push_back(arena_);
 }
 
+bool ColumnAggregateFunction::convertion(MutableColumnPtr *res_) const
```
Comment (KochetovNicolai) [inline dbms/src/Columns/ColumnAggregateFunction.cpp]: Name is not good (difficult to understand, what's happening). Also comment to function is needed.

File: dbms/src/Columns/ColumnAggregateFunction.cpp [hunk 12]
```diff
@@ -65,23 +84,59 @@ MutableColumnPtr ColumnAggregateFunction::convertToValues() const
...
 
     return res;
 }
 
+MutableColumnPtr ColumnAggregateFunction::predictValues(Block & block, const ColumnNumbers & arguments) const
+{
+    MutableColumnPtr res;
+    if (convertion(&res))
+    {
+        return res;
```
Comment (KochetovNicolai) [inline dbms/src/Columns/ColumnAggregateFunction.cpp]: Do we need to return here?

File: dbms/src/Columns/ColumnAggregateFunction.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply dbms/src/Columns/ColumnAggregateFunction.cpp]: Agree. Function should know it's type.

**Review by KochetovNicolai — CHANGES_REQUESTED** [review]

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h (hunk 5 above)
Comment (Quid37) [reply dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: This condition is checked in predictResultInto (line ~520)

File: dbms/src/Columns/ColumnAggregateFunction.cpp (hunk 1 above)
Comment (Quid37) [reply dbms/src/Columns/ColumnAggregateFunction.cpp]: I will add `predictValues()` to base class `IAggregateFunction`

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.cpp [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+    for (size_t i = 0; i < argument_types.size(); ++i)
+    {
+        if (!WhichDataType(argument_types[i]).isFloat64())
+            throw Exception("Illegal type " + argument_types[i]->getName() + " of argument " + std::to_string(i) + "for aggregate function " + name, ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+    }
+
+    Float64 learning_rate = Float64(0.01);
+    Float64 l2_reg_coef = Float64(0.01);
+    UInt32 batch_size = 1;
```
[the remark this answers is not in the document]
Comment (Quid37) [reply dbms/src/AggregateFunctions/AggregateFunctionMLMethod.cpp]: With batch_size = 1 it is simple and purely stochastic. But we can pick another value

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h (hunk 5 above)
Comment (Quid37) [reply dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: This condition is checked in `predictValue()` (line ~520)

File: dbms/src/Columns/ColumnAggregateFunction.cpp (hunk 11 above)
Comment (Quid37) [reply dbms/src/Columns/ColumnAggregateFunction.cpp]: This function is used only in two function below and helps to avoid repetition. Have no idea of a better naming

File: dbms/src/Columns/ColumnAggregateFunction.cpp (hunk 12 above)
Comment (Quid37) [reply dbms/src/Columns/ColumnAggregateFunction.cpp]: No. fixed

File: dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h (hunk 5 above)
Comment (KochetovNicolai) [reply dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h]: Anyway it's better to check in case you change invariant and forget to fix some code.

File: dbms/src/Columns/ColumnAggregateFunction.cpp (hunk 11 above)
Comment (KochetovNicolai) [reply dbms/src/Columns/ColumnAggregateFunction.cpp]: Maybe `tryFinalizeAggregateFunction`

**Review by KochetovNicolai — APPROVED** [review]

File: dbms/tests/queries/0_stateless/00946_ml_test.sql [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
...028633990455,-1.8894018386217877,-1.618648166768315,-1.3414432310256739,-1.0999748867047898),(0.6276564089897139,0.5922213975408968,0.7988286239230257,-0.3977688081309026,-0.5042604443804907,0.690219702968213,0.909111664354187,0.82,-1.8157748668826532,-1.7172770886910105,-0.952028633990455,-1.8894018386217877,-1.618648166768315,-1.3414432310256739,-1.0999748867047898),(-0.3322630208172874,-0.5625278455750569,-0.0766000050337147,0.5966532121963541,1.167321139201246,0.08575800359813185,-1.0999748867047898,0.73,-1.8157748668826532,-1.7172770886910105,-0.952028633990455,-1.8894018386217877,-1.618648166768315,-1.3414432310256739,-1.0999748867047898),(0.1913293954410769,-0.2325994903990701,-0.0766000050337147,0.09944220203272576,-1.0614543055744028,-0.45154128473083044,0.909111664354187,0.74,-1.8157748668826532,-1.7172770886910105,-0.952028633990455,-1.8894018386217877,-1.618648166768315,-1.3414432310256739,-1.0999748867047898),(-0.41952842352701486,-0.06763531281107672,-0.952028633990455,-0.8949798182945309,0.052933416813421515,-0.5522849012925116,-1.0999748867047898,0.69,-1.8157748668826532,-1.7172770886910105,-0.952028633990455,-1.8894018386217877,-1.618648166768315,-1.341443231025673
```
Comment (alexey-milovidov) [inline dbms/tests/queries/0_stateless/00946_ml_test.sql]: Aggregate function must be named in camelCase, not CamelCase.

File: dbms/src/Columns/ColumnAggregateFunction.h [hunk 15]
```diff
@@ -117,6 +118,9 @@ class ColumnAggregateFunction final : public COWHelper<IColumn, ColumnAggregateF
     std::string getName() const override { return "AggregateFunction(" + func->getName() + ")"; }
     const char * getFamilyName() const override { return "AggregateFunction"; }
 
+    bool tryFinalizeAggregateFunction(MutableColumnPtr* res_) const;
```
Comment (alexey-milovidov) [inline dbms/src/Columns/ColumnAggregateFunction.h]: Missing comment.

File: dbms/src/Columns/ColumnAggregateFunction.cpp [hunk 16]
```diff
@@ -33,6 +35,25 @@ void ColumnAggregateFunction::addArena(ArenaPtr arena_)
     arenas.push_back(arena_);
 }
 
+/// This function is used in convertToValues() and predictValues()
+/// and is written here to avoid repetitions
+bool ColumnAggregateFunction::tryFinalizeAggregateFunction(MutableColumnPtr *res_) const
+{
+    if (const AggregateFunctionState *function_state = typeid_cast<const AggregateFunctionState *>(func.get()))
```
Comment (alexey-milovidov) [inline dbms/src/Columns/ColumnAggregateFunction.cpp]: Whitespace.

File: dbms/src/AggregateFunctions/IAggregateFunction.h [hunk 17]
```diff
@@ -92,6 +94,13 @@ class IAggregateFunction
     /// Inserts results into a column.
     virtual void insertResultInto(ConstAggregateDataPtr place, IColumn & to) const = 0;
 
+    /// This function is used for machine learning methods
+    virtual void predictValues(ConstAggregateDataPtr /* place */, IColumn & /*to*/,
```
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/IAggregateFunction.h]: Questionable.

File: dbms/src/Columns/ColumnAggregateFunction.cpp [hunk 18]
```diff
@@ -33,6 +35,25 @@ void ColumnAggregateFunction::addArena(ArenaPtr arena_)
     arenas.push_back(arena_);
 }
 
+/// This function is used in convertToValues() and predictValues()
+/// and is written here to avoid repetitions
+bool ColumnAggregateFunction::tryFinalizeAggregateFunction(MutableColumnPtr *res_) const
```
Comment (alexey-milovidov) [inline dbms/src/Columns/ColumnAggregateFunction.cpp]: Why not return by value?

Comment (alexey-milovidov) [inline dbms/src/Columns/ColumnAggregateFunction.cpp]: Bad method name.
